Add explanation of two link styles (#289)
grafana/writers-toolkit: docs/sources/write/markdown-guide/index.md

For information about creating links between topics inside and outside of a Grafana Labs repository, refer to [Links and cross references]({{< relref "../references" >}}).

> There are two forms of links in Markdown: inline and reference-style.

Reviewer: OMG, did this really say "curly brackets' when it was parentheses?

Inline links define the link text and destination inline.
The following snippet demonstrates an inline link with the text "Link text to display" and the destination https://example.com.

```markdown
[Link text to display](https://example.com)
```

> Reference-style links use labels to join link text and destinations.
> They allow a destination to be defined once for a label, and that label to be reused multiple times in the document.

Reviewer: ```suggestion
When you create a reference-style link, you use define your link text and then use a label to reference the link destination that is defined somewhere else in the document, usually at the end of the file.
Reference-style links let you define a link destination once, and then reuse the label multiple times in the document.
```

Reviewer 2: Line 82... there's an extra word in the suggested text

> When you create a reference-style link, you **use** define your link text and then use a

Reviewer: Fixed, thanks @clayton-cornell !

> The following snippet demonstrates a reference-style link with the text "Link text to display", the destination https://example.com, and uses the label "label".

Reviewer 3: Do we want to explain where on the page the label can live, or is this unnecessary?

Author: How about:

```suggestion
The following snippet demonstrates a reference-style link with the text "Link text to display", the destination https://example.com, and uses the label "label".

The destination can be defined anywhere in the page but is typically put at the bottom of the page to imitate footnotes.
```

Reviewer: I think it's a good idea to establish a standard for defining labels at the end of the file.

Reviewer 2: Do we want'need to state that the label is "optional" and the link text can be used as the label in the definition at the bottom of the page... or should we not allow/encourage this? and opt for the full `[text][label]` syntax even when text and label may be identical?

Author: It's a really good question and hits at the core purpose of this documentation.

Are we teaching people GitHub Flavored Markdown (GFM) or are we teaching people what is common in Markdown at Grafana?

I'm not sure what the answer is right now and I think it would make sense to add whatever we decide to the front matter to remind ourselves in the future.

In terms of this PR, if we are teaching GFM, I think I will need to mention each of the additional cases that omit link labels. If we are instead teaching common Grafana practice, we could skip that.

Maybe we want both the comprehensive information about GFM and the Grafana Labs recommendations?

Reviewer: I thought the whole point of the Writer's Toolkit was to document Grafana practice?  There are other places on the Internet to learn GitHub Flavored Markedown.

Author: I think I've convinced myself in writing that comment that we should explain all the quirks of Markdown *and* include our best guidance.

I'll find some time to update this PR with an explanation of omitting link labels.

Author: > I thought the whole point of the Writer's Toolkit was to document Grafana practice? There are other places on the Internet to learn GitHub Flavored Markedown.

I think understanding GFM might be key to achieving the stated purpose of Writers' Toolkit:

> By creating Writers’ Toolkit, we want to empower everyone to create technical documentation that is consistent and applies the voice and tone we use at Grafana Labs so that our documentation is truly a reflection of who we are.

The teaching of GFM empowers contributors, the explanation of best practices supports consistency.

Reviewer: But we standardized on Commonmark, not GitHub flavored markdown.  And that's what we have documented in the Writer's Toolkit.

"Hugo uses a Markdown parser named Goldmark, which supports the CommonMark flavor of Markdown including some extended features. For more information, see the [CommonMark specification](https://spec.commonmark.org/), and a [quick reference guide](https://commonmark.org/help/) for CommonMark."

This was one of the first questions I asked when I started working here, which markdown standard we were following.

Author: Oh right, I forgot that. I guess the same applies, just `s/GFM/CommonMark/g`.

I think there are some extensions that are not strictly CommonMark. I'm thinking specifically of heading identifier generation which I think might be GFM inspired. A quick search in the CommonMark spec didn't bring up anything.

But yeah, that spec and the quick reference is really useful, perhaps I should start linking people to that instead of our documentation.

For information about creating links between topics inside and outside of a Grafana Labs repository, refer to [Links and cross references]({{< relref "../references" >}}).

There are two forms of links in Markdown: inline and reference-style.

When you create an inline link, you define the link text and destination in the same location in the document.
The following snippet demonstrates an inline link with the text "Link text to display" and the destination https://example.com.

```markdown
[Link text to display](https://example.com)
```

When you create a reference-style link, you define your link text and then use a label to reference the link destination that is defined somewhere else in the document, usually at the end of the file.
Reference-style links let you define a link destination once, and then reuse the label multiple times in the document.

The following snippet demonstrates a reference-style link with the text "Link text to display", the destination https://example.com, and uses the label "label".

> The destination can be defined anywhere in the page but is typically put at the bottom of the page to imitate footnotes.

Author: ```suggestion
```
